packages/server/src/service-worker-manager.ts

```diff
@@ -89,47 +99,67 @@ export class ServiceWorkerManager implements IServiceWorkerManager {
    * Initialize the Service Worker
    */
   private async _initialize(workerUrl: string): Promise<void> {
-    const { serviceWorker } = navigator;
-
-    let registration: ServiceWorkerRegistration | null = null;
+    let registration: ServiceWorkerRegistration | undefined = undefined;
 
-    if (!serviceWorker) {
+    if (!navigator.serviceWorker) {
       console.warn('ServiceWorkers not supported in this browser');
       this._ready.reject(void 0);
       return;
-    } else if (serviceWorker.controller) {
-      const scriptURL = serviceWorker.controller.scriptURL;
-      await this._unregisterOldServiceWorkers(scriptURL);
+    }
 
-      registration = (await serviceWorker.getRegistration(scriptURL)) || null;
-      // eslint-disable-next-line no-console
-      console.info('JupyterLite ServiceWorker was already registered');
+    if (
+      await this._serviceWorkerIsOutdated(navigator.serviceWorker.controller?.scriptURL)
+    ) {
+      const registrations = await navigator.serviceWorker.getRegistrations();
+      await Promise.all(registrations.map((registration) => registration.unregister()));
     }
 
-    if (!registration && serviceWorker) {
-      try {
-        // eslint-disable-next-line no-console
-        console.info('Registering new JupyterLite ServiceWorker', workerUrl);
-        registration = await serviceWorker.register(workerUrl);
-        // eslint-disable-next-line no-console
-        console.info('JupyterLite ServiceWorker was sucessfully registered');
-      } catch (err: any) {
-        console.warn(err);
-        console.warn(
-          `JupyterLite ServiceWorker registration unexpectedly failed: ${err}`,
+    // Check if service worker is active, if not register and wait for active state
+    registration = await navigator.serviceWorker.getRegistration();
+
+    try {
+      if (!registration || !registration.active) {
+        await navigator.serviceWorker.register(workerUrl);
+        await navigator.serviceWorker.ready;
+        if (!navigator.serviceWorker.controller) {
+          // This happens upon hard refresh, the service worker doesn't take over the page, we need to reload the page
+          window.location.reload();
```

Comment:
In terms of UX, a page reload without any confirmation from the user can be quite disruptive. Even though the notebook dirty state tracker should prevent that if some changes have not been saved yet.

Maybe a dialog or notification to inform the user, with a button to do the reload, could help.

Reply:
This would be very annoying. 

This page reload is actually not noticeable to the user, since it directly reload the page without the user seeing the UI yet.

On chrome, a page reload is required after a hard refresh if there are other tabs open, otherwise the service worker doesn't control the newly refreshed page.